### PR TITLE
Allow sound instances to play custom audio streams

### PR DESCRIPTION
**Pull request by SquidDev**
There's a couple of mods out there which play dynamically generated sounds, rather than reading from an ogg file. For instance, [Etched](https://github.com/MoonflowerTeam/etched) allows streaming audio from an arbitrary URL and [CC: Restitched](https://github.com/cc-tweaked/cc-restitched) allows players to generate music data and stream it to the client.

Minecraft's sound engine supports custom audio streams. However, as there is no API exposed for this, mods must use mixins. Unfortunately the sound code is fairly complex and, due to its multi-threaded and callback-based nature, awkward to mixin to. For instance (note, most of these examples are using MojMap):

 - [Etched uses `@Redirect`](https://github.com/MoonflowerTeam/etched/blob/8d459ef623a869f6a87cd57f4d634127dd3db7b2/fabric/src/main/java/gg/moonflower/etched/core/mixin/fabric/client/SoundEngineMixin.java#L33-L38). This is the most "correct" method (and is what this PR also does), but is incompatible with any other mod.
 - [CC: Restitched `@Inject`s into the `play` method and a callback lambda](https://github.com/cc-tweaked/cc-restitched/blob/55625c72516fda8290820ec57fdbc994bf06dc1c/src/main/java/dan200/computercraft/fabric/mixin/MixinSoundEngine.java#L29-L50). This code is incorrect, due to the multi-threaded nature of the sound code (though works well enough in practice).
 - Some mods reimplement some (or all) of the `play` method (i.e. [Infinidisc](https://github.com/DenimsMods/Infinidisc/blob/045dd7190b16c51795ee1ee6abaf162d3d6ca10c/src/main/java/denimred/infinidisc/mixin/SoundEngineMixin.java#L53), [music-box](https://github.com/CmdrJane/music-box/blob/c008602ae615af66882c61a4e81e42561c16cbdc/src/main/java/ru/aiefu/musicbox/mixin/SoundEngineMixin.java#L60)).

Given several mods want to provide custom `AudioStream`s, but doing so in a compatible way is tricky to get right, it makes sense for Fabric to provide an API everyone can hook into.

This PR adds a interface `FabricSoundInstance`, injected into `SoundInstance`. This provides a new method `getAudioStream`, which can be overridden to return a custom audio stream. The default implementation implements the current behaviour of just reading an `.ogg` file from resource packs.

## Further notes
 - The obvious alternative here would be to add an event, instead of providing a new interface. The current approach felt more natural, as whenever you want this functionality, you're going to be sub-classing `SoundInstance` anyway. Happy to go for either.

 - This does still require `SoundInstance.getSound` to resolve to an actual file, even if that file isn't read by the audio stream. This is definitely a little clunky; I'm just not sure if "fixing" this is worth the additional implementation cost. Thoughts welcome!

 - I've added a new module `fabric-sound-api-v1` as nowhere else felt natural. Obviously let me know if I've got anything wrong with adding the new module :).

 - It appears that injected interfaces only work when registered in the main source set. I'm not sure if this is intentional or not?

### CODE REVIEW

**Comment by modmuss50** [comment]
> It appears that injected interfaces only work when registered in the main source set.

I dont think thats intentional, most likely something I can fix in loom. Ideally client only modules should have everything in the client sourceset as this will prevent the need to split the jar in a mod dev env. I will take a look into this.

Had a quick look at the PR, it looks great at first glance. I will take a deeper look soon. 👍

**Comment by modmuss50** [comment]
This should be done by default, something I can fix in a future loom version, for now you should be able to add the following to the sub projects build.gradle:

```gradle
loom {
	interfaceInjection {
		interfaceInjectionSourceSets.add sourceSets.client
	}
}
```

File: fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/api/client/sound/v1/FabricSoundInstance.java
Comment (apple502j) [inline fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/api/client/sound/v1/FabricSoundInstance.java]:
```suggestion
	 * Loads the audio stream for this sound.
```

File: fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/api/client/sound/v1/FabricSoundInstance.java
Comment (apple502j) [inline fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/api/client/sound/v1/FabricSoundInstance.java]:
```suggestion
	 * <p>By default, this will load {@code .ogg} files from active resource packs. It may be overridden to provide a
```

File: fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/api/client/sound/v1/FabricSoundInstance.java
Comment (apple502j) [inline fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/api/client/sound/v1/FabricSoundInstance.java:83]:
```suggestion
	 * @param id              The resolved sound ID, equal to {@link SoundInstance#getSound()}'s.
```

File: fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/api/client/sound/v1/FabricSoundInstance.java
Comment (apple502j) [inline fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/api/client/sound/v1/FabricSoundInstance.java]:
```suggestion
	 * @return the loaded audio stream
```

File: fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/mixin/client/sound/SoundSystemMixin.java
Comment (apple502j) [inline fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/mixin/client/sound/SoundSystemMixin.java]:
```suggestion
	private CompletableFuture<?> getStream(SoundLoader loader, Identifier id, boolean looping, SoundInstance sound) {
		return sound.getAudioStream(loader, id, looping);
```

File: fabric-sound-api-v1/src/client/resources/fabric-sound-api-v1.mixins.json
Comment (apple502j) [inline fabric-sound-api-v1/src/client/resources/fabric-sound-api-v1.mixins.json]:
```suggestion
  "compatibilityLevel": "JAVA_17",
```

File: fabric-sound-api-v1/src/testmod/java/net/fabricmc/fabric/test/sound/client/SineSound.java
Comment (apple502j) [inline fabric-sound-api-v1/src/testmod/java/net/fabricmc/fabric/test/sound/client/SineSound.java]:
```suggestion
	protected SineSound(Vec3d pos) {
		super(new Identifier(ClientSoundTest.MOD_ID, "sine_wave"), SoundCategory.BLOCKS, SoundInstance.createRandom());
		this.x = pos.x;
		this.y = pos.y;
		this.z = pos.z;
```

File: fabric-sound-api-v1/src/testmod/java/net/fabricmc/fabric/test/sound/client/SineSound.java
Comment (apple502j) [inline fabric-sound-api-v1/src/testmod/java/net/fabricmc/fabric/test/sound/client/SineSound.java]:
```suggestion
	public CompletableFuture<AudioStream> getAudioStream(SoundLoader loader, Identifier id, boolean repeatInstantly) {
```

File: fabric-sound-api-v1/src/testmod/java/net/fabricmc/fabric/test/sound/client/SineStream.java
Comment (apple502j) [inline fabric-sound-api-v1/src/testmod/java/net/fabricmc/fabric/test/sound/client/SineStream.java:21]: Just making sure, this won't cause a mess in macOS right?

**Review by apple502j — CHANGES_REQUESTED** [review]
Also why are you shipping the empty ogg farm as part of the API (and not testmod)?

File: fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/api/client/sound/v1/FabricSoundInstance.java
Comment (SquidDev) [reply fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/api/client/sound/v1/FabricSoundInstance.java:83]: `SoundInstance.getSound()` returns a `Sound`, not an `Identifier`, so I don't think this makes much sense.

**Comment by SquidDev** [comment]
> Also why are you shipping the empty ogg farm as part of the API (and not testmod)?

As every mod that provides a custom audio stream needs to reference a dummy sound it made sense to provide it in the API, rather than every mod having to bundle their own - much like fabric-gametest-api-v1's empty structure.

File: fabric-sound-api-v1/src/testmod/java/net/fabricmc/fabric/test/sound/client/SineStream.java
Comment (SquidDev) [reply fabric-sound-api-v1/src/testmod/java/net/fabricmc/fabric/test/sound/client/SineStream.java:21]: I don't have a Mac to test, but I'd be surprised if it doesn't work. It's just part of the standard JRE (`java.desktop` module, same as Swing) and given that vanilla MC uses it, I feel we'd have seen issues much earlier.

Comment (modmuss50) [reply fabric-sound-api-v1/src/testmod/java/net/fabricmc/fabric/test/sound/client/SineStream.java:21]: Just tested it on my ARM mac and this works as expected. I dont think there is any concern here as its not an AWT/swing class. The vanilla usage proves this is a none issue.

File: fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/api/client/sound/v1/FabricSoundInstance.java
Comment (apple502j) [reply fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/api/client/sound/v1/FabricSoundInstance.java:83]: Well, maybe `'s ID`.

**Comment by apple502j** [comment]
@SquidDev Then the existence should be documented somewhere, along with `Identifier` of the empty sound exposed as a field.

File: fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/api/client/sound/v1/FabricSoundInstance.java
Comment (SquidDev) [reply fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/api/client/sound/v1/FabricSoundInstance.java:83]: I went for `location` because that's the name of the field & getter `Sound`. ID is _fine_, but I don't think it indicates exactly what is being referred to here as well.

Comment (SquidDev) [reply fabric-sound-api-v1/src/client/java/net/fabricmc/fabric/api/client/sound/v1/FabricSoundInstance.java:83]: I think part of the problem here is that it's important to distinguish between the id of the `SoundInstance` (which corresponds to the id in `sounds.json` and `SoundEvent`), and the id of the actual `Sound` being played (which corresponds to a path in a resource pack). Separating the two by having different nomenclature for both feels helpful.

**Comment by SquidDev** [comment]
> Then the existence should be documented somewhere, along with Identifier of the empty sound exposed as a field.

:+1:  Is it acceptable just to add it as a constant to `FabricSoundInstance`? Not sure there's a more sensible location until #2503 is resolved.

**Review by Technici4n — APPROVED** [review]
Looks good to me!

**Comment by modmuss50** [comment]
> Is it acceptable just to add it as a constant to FabricSoundInstance?

Yes, I think that would be a good idea 👍

**Review by modmuss50 — APPROVED** [review]
Looks good to me, just need to document the existence of the empty sound, and why is required.

**Comment by SquidDev** [comment]
I felt it was probably worth while adding an example, as the need for `sounds.json` is a little awkward and thus worth being explicit in. Happy to revert if it's Too Much :).

**Comment by SquidDev** [comment]
Hah, I should know better than to commit without running checkstyle!

I don't know if there's any interest in adding a pre-commit hook which runs it automatically? Or is that just going to get too irritatingly slow for people?

**Comment by Technici4n** [comment]
More documentation is always welcome. :)

Not sure about the pre-commit hook as I've never used them myself, but it looks like it might end up being rather annoying ? We always wait for CI to pass before merging anything.

**Comment by Juuxel** [comment]
IMO it's too slow since running `gradlew checkstyleMain checkstyleTestmod` (or even just the main one) builds all modules and that takes a while even with `--parallel`.

**Review by modmuss50 — APPROVED** [review]

**Comment by modmuss50** [comment]
I agree, spotless would also need to be ran making commits very slow. I would reccomend using the checkstyle plugin in your IDE to make it clear when you have an issue.

Added to last call, thanks 👍

**Comment by SquidDev** [comment]
Thank you everyone :).

**Pull request merged**
